### PR TITLE
MM-10007 Send an admin and regular WS events when a user is updated

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
Send an admin and regular WS events when a user is updated.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-10007

### CODE REVIEW

File: app/web_conn.go [hunk 1]
```diff
@@ -287,6 +287,14 @@ func (webCon *WebConn) ShouldSendEvent(msg *model.WebSocketEvent) bool {
 		return false
 	}
 
+	if msg.Broadcast.OmitAdmins && webCon.App.RolesGrantPermission(webCon.GetSession().GetUserRoles(), model.PERMISSION_MANAGE_SYSTEM.Id) {
+		return false
+	}
+
+	if msg.Broadcast.OmitNonAdmins && !webCon.App.RolesGrantPermission(webCon.GetSession().GetUserRoles(), model.PERMISSION_MANAGE_SYSTEM.Id) {
```
Comment (jwilander) [inline app/web_conn.go]: These should, in theory, be expensive checks as, now and in the future, we should already have the user's roles, the roles themselves and permissions either locally in a memory cache or hard-coded.

**Comment by jwilander** [comment]
Alright, this isn't quite complete because I want to get some thoughts on it before I go any further to  add more unit tests and do a load test.

@crspeller, I chatted with @lieut-data earlier today and while we think it's possible to find a way to send nullable user fields for the ones that might get sanitized out, that really only fixes the case where one of these fields didn't actually change. If they were to change, then the admin would have out-of-date data on the client until they did a re-fetch of the user. It was also becoming more complex to present those fields in a nullable way cleanly since the User model isn't using pointers for these fields. Jesse and I brainstormed some ways around that but none of them were very nice. So instead, we opted to go with something similar to what we had discussed before. Thoughts?

@grundleborg, the issue being solved is described in the ticket but the short version is we need a way to separately sanitize WS events for admins and regular users. Now I realize this might step on advanced permissions toes a little bit by introducing more notions around "IsAdmin" and "IsNotAdmin". If you have any suggestions about how I can go about this in a more advanced-permissions-friendly way I'd love to hear and implement one of them

@lieut-data, please look as well and share any thoughts

File: app/web_conn.go (hunk 1 above)
Comment (crspeller) [reply app/web_conn.go]: Can we at least not run the same call out to the permissions system twice in a row here.

**Comment by crspeller** [comment]
I think the case where someone with system admin permissions needs to refresh to get someone's updated email might be acceptable if the alternative is having problems with the permissions system. 

Ultimately I think @grundleborg should have the final say on that.

File: app/web_conn.go (hunk 1 above)
Comment (jwilander) [reply app/web_conn.go]: I can update it but it's only ever going to run twice if both `OmitAdmins` and `OmitNonAdmins` are set which /shouldn't/ ever happen. I can also pull it out completely and just store it on the on the webconn struct

Comment (jwilander) [reply app/web_conn.go]: Awesome, thanks George. I'll update the variable names to something more appropriate

Comment (crspeller) [reply app/web_conn.go]: Ya, thanks George!

**Comment by jwilander** [comment]
@crspeller can you take another look and approve if you're OK with it?

**Pull request merged**